README.md
GTimeUtils
=========

**A simple stopwatch and timer, written in GTK3.**

Usage
-----

A small word about the usage of, for now, Gtimer. Gtimer can take four (five if you count the help message) command line arguments:

	-s, --seconds     Specify seconds to count down from
	-m, --minutes     Specify minutes to count down from
	-u, --hours       Specify hours to count down from
	-t, --text        Set an alternative notification text
	-r, --run         Immediately start the countdown

Installation
------------

GTimeUtils is written in GTK 3.x, so you need GTK3's headers with the libnotify and libcanberra headers installed and ready to go. To
build, simply run these commands:

	$ ./autogen.sh
	$ make
	# make clean install

Translations
-----------

You can help translate GTimeUtils in your language!
To do so, simply follow these steps:

	$ cd po
	$ intltool-update --pot
	$ mv gtimeutils.pot <language code>.po

Where `<language code>` is, obviously, the code of your language (e.g. `nl` for Dutch, `fr` for French, `en` for English...)
Edit the `LINGUAS` file and add your language code. Please keep the list alphabetically.
Lastly, open the .po file you just generated and translate all the strings. Don't forget to fill in the information in the header!

Bugs
----

For any bug or request [fill an issue][bug] on [GitHub][ghp].

  [bug]: https://github.com/Unia/gtimeutils/issues
  [ghp]: https://github.com/Unia/gtimeutils

License
-------

**GTimeUtils** is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.

**GTimeUtils** is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with this program.  If not, see <http://www.gnu.org/licenses/>.

**Copyright © 2013** Jente Hidskes <jthidskes@outlook.com>


ChangeLog
 2013-06-10: v0.1: Imported GTimeUtils into Anjuta for easier management
                   Removed GTK2 support



src/gstopwatch.c
/* -*- Mode: C; indent-tabs-mode: t; c-basic-offset: 4; tab-width: 4 -*-  */
/*
 * gstopwatch.c
 * Copyright (C) 2013 Jente Hidskes <jthidskes@outlook.com>
 * 
 * gtimeutils is free software: you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the
 * Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * gtimeutils is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License along
 * with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <config.h>
#include <gtk/gtk.h>
#include <glib/gi18n.h>

enum {
    STARTED,
    PAUSED,
    STOPPED
};

enum {
    N_LAP,
    TIME,
    N_COLUMNS
};

GTimer *stopwatch;
gchar output[100];
gint state = STOPPED, laps = 0, hours = 0, minutes = 0;
gdouble seconds = 0;
GtkWidget *stopwatch_display, *button_stopwatch, *button_funcs, *tree;
GtkListStore *liststore;
GtkTreeSelection *selection;
GtkTreeIter selection_iter, iter;

void counter (gboolean counting) {
	gchar *markup;

	if(counting) {
		hours = seconds / 3600;
		seconds -= 3600 * hours;
		minutes = seconds / 60;
		seconds -= 60 * minutes;
		sprintf(output, "%02d:%02d:%.2f", hours, minutes, seconds);
	} else
		sprintf(output, "00:00:0,00");

	gtk_label_set_text (GTK_LABEL (stopwatch_display), output);
	markup = g_markup_printf_escaped ("<span font=\"48\" weight=\"heavy\"><tt>%s</tt></span>", output);
	gtk_label_set_markup (GTK_LABEL (stopwatch_display), markup);
	g_free (markup);
}

gboolean stopwatch_function (void) {
	gulong gulong;

	if(state == STARTED) {
		seconds = g_timer_elapsed (stopwatch, &gulong);
		counter (TRUE);
	} else if(state == STOPPED)
		counter (FALSE);
	return TRUE;
}

void add_lap (void) {
	laps++;

	gtk_list_store_append (GTK_LIST_STORE (liststore), &iter);
	gtk_list_store_set (GTK_LIST_STORE (liststore), &iter, N_LAP, laps, TIME, output, -1);
	gtk_tree_selection_select_iter (gtk_tree_view_get_selection(GTK_TREE_VIEW(tree)), &iter);
}

void on_stopwatch_button_clicked (void) {
	if(state == STOPPED) {
		gtk_button_set_label (GTK_BUTTON (button_stopwatch), _("Stop"));
		gtk_widget_set_sensitive (GTK_WIDGET (button_funcs), TRUE);
		gtk_button_set_label (GTK_BUTTON (button_funcs), _("Lap"));
		g_timer_start (stopwatch);
		state = STARTED;
	} else if(state == PAUSED) {
		gtk_button_set_label (GTK_BUTTON (button_stopwatch), _("Stop"));
		gtk_button_set_label (GTK_BUTTON (button_funcs), _("Lap"));
		g_timer_continue (stopwatch);
		state = STARTED;
	} else if(state == STARTED) {
		gtk_button_set_label (GTK_BUTTON (button_stopwatch), _("Continue"));
		gtk_widget_set_sensitive (GTK_WIDGET (button_funcs), TRUE);
		gtk_button_set_label (GTK_BUTTON (button_funcs), _("Reset"));
		g_timer_stop (stopwatch);
		state = PAUSED;
	}
}

void on_funcs_button_clicked (void) {
	if(state == STARTED)
		add_lap();
	else if(state == PAUSED) {
		g_timer_start (stopwatch);
		g_timer_stop (stopwatch);
		laps = 0;
		state = STOPPED;
		gtk_widget_set_sensitive (GTK_WIDGET (button_funcs), FALSE);
		gtk_button_set_label (GTK_BUTTON (button_funcs), _("Reset"));
		gtk_button_set_label (GTK_BUTTON (button_stopwatch), _("Start"));
		gtk_list_store_clear (GTK_LIST_STORE (liststore));
	}
}

int main (void) {
 	GtkWidget *window, *vbox, *hbox, *scroll;
	GtkTreeViewColumn *column;
	GtkCellRenderer *renderer;

#ifdef ENABLE_NLS
	bindtextdomain (GETTEXT_PACKAGE, PACKAGE_LOCALE_DIR);
	bind_textdomain_codeset (GETTEXT_PACKAGE, "UTF-8");
	textdomain (GETTEXT_PACKAGE);
#endif
	
	gtk_init (NULL, NULL);

	vbox = gtk_box_new (GTK_ORIENTATION_VERTICAL, 5);
	hbox = gtk_box_new (GTK_ORIENTATION_HORIZONTAL, 5);
	stopwatch_display = gtk_label_new ("");
	button_stopwatch = gtk_button_new_with_label (_("Start"));
	button_funcs = gtk_button_new_with_label (_("Reset"));
	gtk_widget_set_sensitive (button_funcs, FALSE);
	scroll = gtk_scrolled_window_new (NULL, NULL);
	gtk_scrolled_window_set_shadow_type (GTK_SCROLLED_WINDOW (scroll), GTK_SHADOW_IN);
	gtk_scrolled_window_set_policy (GTK_SCROLLED_WINDOW (scroll), GTK_POLICY_AUTOMATIC, GTK_POLICY_AUTOMATIC);

	liststore = gtk_list_store_new (N_COLUMNS, G_TYPE_INT, G_TYPE_STRING);
	tree = gtk_tree_view_new_with_model (GTK_TREE_MODEL (liststore));

	column = gtk_tree_view_column_new();
	gtk_tree_view_column_set_title (column, _("Lap"));

	renderer = gtk_cell_renderer_text_new();
	gtk_tree_view_column_pack_start (column, renderer, TRUE);
	gtk_tree_view_column_set_attributes (column, renderer, "text", N_LAP, NULL);
	gtk_tree_view_column_set_expand (column, TRUE);
	gtk_tree_view_append_column (GTK_TREE_VIEW (tree), column);

	renderer = gtk_cell_renderer_text_new();
	column = gtk_tree_view_column_new_with_attributes (_("Time"), renderer, "text", TIME, NULL);
	gtk_tree_view_column_set_expand (column, TRUE);
	gtk_tree_view_append_column (GTK_TREE_VIEW (tree), column);

	selection = gtk_tree_view_get_selection (GTK_TREE_VIEW (tree));
	gtk_tree_selection_set_mode (selection, GTK_SELECTION_SINGLE);
	gtk_tree_view_set_enable_search (GTK_TREE_VIEW (tree), FALSE);

	gtk_box_pack_start (GTK_BOX (vbox), stopwatch_display, FALSE, TRUE, 5);
	gtk_box_pack_start (GTK_BOX (hbox), button_stopwatch, TRUE, TRUE, 5);
	gtk_box_pack_start (GTK_BOX (hbox), button_funcs, TRUE, TRUE, 5);
	gtk_container_add (GTK_CONTAINER (vbox), hbox);
	gtk_container_add (GTK_CONTAINER (scroll), tree);
	gtk_box_pack_start (GTK_BOX (vbox), scroll, TRUE, TRUE, 0);

	window = gtk_window_new (GTK_WINDOW_TOPLEVEL);
	gtk_window_set_title (GTK_WINDOW (window), "Gstopwatch");
	gtk_container_add (GTK_CONTAINER (window), vbox);
	gtk_widget_show_all (window);

	stopwatch = g_timer_new();
	g_timer_stop (stopwatch);

	g_timeout_add_full (G_PRIORITY_HIGH, 50, (GSourceFunc) stopwatch_function, NULL, NULL);
	g_signal_connect (G_OBJECT (window), "destroy", G_CALLBACK (gtk_main_quit), NULL);
	g_signal_connect (button_stopwatch, "clicked", G_CALLBACK (on_stopwatch_button_clicked), NULL);
	g_signal_connect (button_funcs, "clicked", G_CALLBACK (on_funcs_button_clicked), NULL);

	gtk_main();
	g_timer_destroy (stopwatch);
	return 0;
}



src/gtimer.c
/* -*- Mode: C; indent-tabs-mode: t; c-basic-offset: 4; tab-width: 4 -*-  */
/*
 * gtimer.c
 * Copyright (C) 2013 Jente Hidskes <jthidskes@outlook.com>
 * 
 * gtimeutils is free software: you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the
 * Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * gtimeutils is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License along
 * with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <config.h>
#include <gtk/gtk.h>
#include <canberra.h>
#include <glib/gi18n.h>
#include <glib/gprintf.h>
#include <libnotify/notify.h>
#include <stdlib.h>

enum {
    STARTED,
    PAUSED,
    STOPPED
};

ca_context *sound;
static gboolean start_on_run = FALSE;
const gchar *entry_text = N_("Notification text");
gint state = STOPPED, hours = 0, minutes = 0, seconds = 0;
GtkWidget *timer_display, *hbox1, *entry, *button_timer, *button_reset, *spin_seconds, *spin_minutes, *spin_hours;

static GOptionEntry entries[] = {
	{ "seconds", 's', 0, G_OPTION_ARG_INT, &seconds, N_("Specify seconds to count down from"), NULL },
	{ "minutes", 'm', 0, G_OPTION_ARG_INT, &minutes, N_("Specify minutes to count down from"), NULL },
	{ "hours", 'u', 0, G_OPTION_ARG_INT, &hours, N_("Specify hours to count down from"), NULL },
	{ "text", 't', 0, G_OPTION_ARG_STRING, &entry_text, N_("Set an alternative notification text"), NULL },
	{ "run", 'r', 0, G_OPTION_ARG_NONE, &start_on_run, N_("Immediately start the countdown"), NULL },
	{ NULL },
};

void reset_display (void) {
	gchar *markup;

	markup = g_markup_printf_escaped ("<span font=\"48\" weight=\"heavy\"><tt>%s</tt></span>", _("HH:MM:SS"));
	gtk_label_set_markup (GTK_LABEL (timer_display), markup);
	g_free (markup);
}

void notify (void) {
	NotifyNotification *notify;
	GError *error_notify = NULL;

	entry_text = gtk_entry_get_text (GTK_ENTRY (entry));
	if(g_strcmp0 (entry_text, "Notification text") != 0)
		notify = notify_notification_new (entry_text, NULL, "clocks");
	else
		notify = notify_notification_new (_("Time is up!"), NULL, "clocks");
	notify_notification_set_category (notify, "GTimeUtils");
	notify_notification_set_urgency (notify, NOTIFY_URGENCY_NORMAL);
	notify_notification_show (notify, &error_notify);
	if(error_notify)
		g_fprintf (stderr, "Can not initialize notification: %s\n", error_notify->message);

	ca_context_play(sound, 0, CA_PROP_APPLICATION_NAME, "Gtimer", CA_PROP_EVENT_ID, "complete-copy", CA_PROP_MEDIA_ROLE, "notification",  CA_PROP_APPLICATION_ICON_NAME, "clocks", CA_PROP_CANBERRA_CACHE_CONTROL, "never", NULL);
}

void counter (void) {
	gchar *markup, output[100];

	seconds = gtk_spin_button_get_value (GTK_SPIN_BUTTON (spin_seconds));
	minutes = gtk_spin_button_get_value (GTK_SPIN_BUTTON (spin_minutes));
	hours = gtk_spin_button_get_value (GTK_SPIN_BUTTON (spin_hours));
	seconds += 60 * minutes;
	seconds += 3600 * hours;

	if(seconds == 0) {
		notify();
		gtk_button_set_label (GTK_BUTTON (button_timer), _("Start"));
		gtk_widget_set_sensitive (button_reset, FALSE);
		gtk_widget_set_sensitive (entry, TRUE);
		state = STOPPED;
		start_on_run = FALSE;
	} else {
		seconds -= 1;
		hours = seconds / 3600;
		seconds -= 3600 * hours;
		minutes = seconds / 60;
		seconds -= 60 * minutes;
		gtk_spin_button_set_value (GTK_SPIN_BUTTON (spin_seconds), seconds);
		gtk_spin_button_set_value (GTK_SPIN_BUTTON (spin_minutes), minutes);
		gtk_spin_button_set_value (GTK_SPIN_BUTTON (spin_hours), hours);
		sprintf (output, "%02d:%02d:%02d", hours, minutes, seconds);

		gtk_label_set_text (GTK_LABEL (timer_display), output);
		markup = g_markup_printf_escaped ("<span font=\"48\" weight=\"heavy\"><tt>%s</tt></span>", output);
		gtk_label_set_markup (GTK_LABEL (timer_display), markup);
		g_free (markup);
	}
}

gboolean timer_function (void) {
	if(state == STARTED || start_on_run)
		counter();
	return TRUE;
}

void on_timer_button_clicked (void) {
	if(state == STOPPED) {
		gtk_button_set_label (GTK_BUTTON (button_timer), _("Stop"));
		gtk_widget_set_sensitive (button_reset, TRUE);
		gtk_widget_set_sensitive (entry, FALSE);
		state = STARTED;
	} else if(state == PAUSED) {
		gtk_button_set_label (GTK_BUTTON (button_timer), _("Stop"));
		state = STARTED;
	} else if(state == STARTED) {
		gtk_button_set_label (GTK_BUTTON (button_timer), _("Continue"));
		state = PAUSED;
	}
}

void on_reset_button_clicked (void) {
	if(state == PAUSED || state == STARTED) {
		state = STOPPED;
		gtk_button_set_label (GTK_BUTTON (button_timer), _("Start"));
		reset_display();
		gtk_spin_button_set_value (GTK_SPIN_BUTTON (spin_seconds), 0);
		gtk_spin_button_set_value (GTK_SPIN_BUTTON (spin_minutes), 0);
		gtk_spin_button_set_value (GTK_SPIN_BUTTON (spin_hours), 0);
		gtk_widget_set_sensitive (button_reset, FALSE);
		gtk_widget_set_sensitive (entry, TRUE);
	}
}

int main (int argc, char *argv[]) {
	GError *error_parsearg = NULL;
	GOptionContext *context;
	GtkWidget *window, *vbox, *hbox2;
	GtkAdjustment *sadj, *madj, *hadj;

#ifdef ENABLE_NLS
	bindtextdomain (GETTEXT_PACKAGE, PACKAGE_LOCALE_DIR);
	bind_textdomain_codeset (GETTEXT_PACKAGE, "UTF-8");
	textdomain (GETTEXT_PACKAGE);
#endif

	gtk_init (&argc, &argv);
	notify_init ("Gtimer");
	ca_context_create (&sound);

	context = g_option_context_new (_("- a simple countdown timer"));
	g_option_context_add_main_entries (context, entries, NULL);
#ifdef ENABLE_NLS
	g_option_context_set_translation_domain (context, GETTEXT_PACKAGE);
#endif
	if (!g_option_context_parse (context, &argc, &argv, &error_parsearg)) {
		g_fprintf (stderr, "%s\n", error_parsearg->message);
		exit(1);
	}

	vbox = gtk_box_new (GTK_ORIENTATION_VERTICAL, 5);
	hbox1 = gtk_box_new (GTK_ORIENTATION_HORIZONTAL, 0);
	hbox2 = gtk_box_new (GTK_ORIENTATION_HORIZONTAL, 0);

	gtk_box_set_spacing (GTK_BOX (hbox2), 5);
	gtk_container_set_border_width (GTK_CONTAINER (vbox), 5);

	timer_display = gtk_label_new (NULL);
	reset_display();
	
	sadj = gtk_adjustment_new (0, 0, 60, 1, 1, 0);
	madj = gtk_adjustment_new (0, 0, 60, 1, 1, 0);
	hadj = gtk_adjustment_new (0, 0, 24, 1, 1, 0);
	spin_seconds = gtk_spin_button_new (sadj, 1, 0);
	spin_minutes = gtk_spin_button_new (madj, 1, 0);
	spin_hours = gtk_spin_button_new (hadj, 1, 0);

	gtk_spin_button_set_wrap (GTK_SPIN_BUTTON (spin_seconds), TRUE);
	gtk_spin_button_set_value (GTK_SPIN_BUTTON (spin_seconds), seconds);
	g_object_set (spin_seconds, "shadow-type", GTK_SHADOW_IN, NULL);
	gtk_spin_button_set_wrap (GTK_SPIN_BUTTON (spin_minutes), TRUE);
	gtk_spin_button_set_value (GTK_SPIN_BUTTON (spin_minutes), minutes);
	g_object_set (spin_minutes, "shadow-type", GTK_SHADOW_IN, NULL);
	gtk_spin_button_set_wrap (GTK_SPIN_BUTTON (spin_hours), TRUE);
	gtk_spin_button_set_value (GTK_SPIN_BUTTON (spin_hours), hours);
	g_object_set (spin_hours, "shadow-type", GTK_SHADOW_IN, NULL);

	entry = gtk_entry_new();
	gtk_entry_set_text (GTK_ENTRY (entry), entry_text);
	
	button_timer = gtk_button_new();
	gtk_button_set_label (GTK_BUTTON (button_timer), _("Start"));
	button_reset = gtk_button_new_with_label (_("Reset"));
	gtk_widget_set_sensitive (button_reset, FALSE);

	gtk_box_pack_start (GTK_BOX (hbox1), spin_hours, TRUE, FALSE, 5);
	gtk_box_pack_start (GTK_BOX (hbox1), spin_minutes, TRUE, FALSE, 5);
	gtk_box_pack_start (GTK_BOX (hbox1), spin_seconds, TRUE, FALSE, 5);
	gtk_box_pack_start (GTK_BOX (vbox), timer_display, FALSE, TRUE, 5);
	gtk_box_pack_start (GTK_BOX (hbox2), button_timer, TRUE, TRUE, 0);
	gtk_box_pack_start (GTK_BOX (hbox2), button_reset, TRUE, TRUE, 0);
	gtk_container_add (GTK_CONTAINER (vbox), hbox1);
	gtk_box_pack_start (GTK_BOX (vbox), entry, FALSE, TRUE, 5);
	gtk_container_add (GTK_CONTAINER (vbox), hbox2);

	window = gtk_window_new (GTK_WINDOW_TOPLEVEL);
	gtk_window_set_title (GTK_WINDOW (window), "Gtimer");
	gtk_container_add (GTK_CONTAINER (window), vbox);
	gtk_widget_show_all (window);

	g_timeout_add_seconds (1, (GSourceFunc) timer_function, NULL);
	g_signal_connect (window, "destroy", G_CALLBACK (gtk_main_quit), NULL);
	g_signal_connect (button_timer, "clicked", G_CALLBACK (on_timer_button_clicked), NULL);
	g_signal_connect (button_reset, "clicked", G_CALLBACK (on_reset_button_clicked), NULL);

	gtk_main();
	notify_uninit();
	ca_context_destroy (sound);
	g_option_context_free (context);
	return 0;
}
